design-proposals/passt-feature-gate-iface-api-deprecation.md

```diff
@@ -0,0 +1,61 @@
+## Overview
+Following the emerging [Kubevirt Network Binding Plugin API](https://docs.google.com/document/d/13fLaqZuM8El_CT5XzxPv-1epKsB1-Spv_jtXveEkkdQ/edit#heading=h.z27rudfpwy44)
+and introduction of Passt network binding plugin [2](https://github.com/kubevirt/kubevirt/pull/10425) [3](https://github.com/kubevirt/kubevirt/pull/10621), 
+it's now possible to remove Passt network binding implementation from Kubevirt core.
+
+This doc outlines deprecation plan for Passt network binding, interface API and  feature gate.
+
+# Motivation
+Removing Passt network binding implementation from Kubevirt core and use Passt network binding plugin instead, 
+enable reduce the maintenance burden.
+Having Passt binding code outside Kubevirt's core may ease contribution process as there is no need to be familiar 
+with [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt) core code.
+
+## Goals
+- Deprecate Passt network interface API.
+- Remove Passt feature gate.
+- Remove Passt network binding from Kubevirt core.
+  
+## Non Goals
+
+## User Stories
+
+## Repos
+[kubevirt/kubevirt](https://github.com/kubevirt/kubevirt):
+- [Passt binding plugin hook sidecar](https://github.com/kubevirt/kubevirt/tree/main/cmd/sidecars/network-passt-binding)
+- [Passt binding plugin CNI](https://github.com/kubevirt/kubevirt/tree/main/cmd/cniplugins/passt-binding)
+
+# Design
+Deprecating Passt feature-gate and interface API and replacing it with Passt binding plugin
+require some changes users should be aware of and prepared for.
+
+The proposed deprecation plan for Passt binding interface API and feature-gate is as follows:
+
+1. Announce passt feature-gate and interface API deprecation in kubevirt-dev mailing list.
+   This will give users some time to prepare for changes in Kubevirt next release v1.2.0. 
+
+2. Kubevirt next release v1.2.0 should have the following changes: 
+   1. [Passt interface API](http://kubevirt.io/api-reference/main/definitions.html#_v1_interfacepasst) 
+      usage shall be blocked by the admission webhook with a message saying binding plugin API should be
+      used instead.
```

Comment:
That is not possible or adviced due to potential upgrade breakage of existing users who work with this binding in its old form.
Kubevirt must assure the users can upgrade without breaking anything (e.g. when the VM is migrated or restarted, it will fail to come up due to this suggested behavior).

It is also unlikely to be possible to do such changes for the next release, but that is open for debate I guess.
For me, it makes sense to give opportunity for users to change their configuration for at least one version. Considering that they may hear about this only when v1.2 is released.

Reply:
What needs to be done first is to add deprecation warnings on fields in the code and as warnings/logs in the runtime.

Then, in the next version the implementation can be removed, leaving the API fields and warnings/logs. The API fields will then be a no-op.

Reply:
The suggestion as I saw it was:
> usage shall be blocked by the admission webhook with a message

So I am saying we cannot block as that will cause existing VM specs to be broken completely.

> Then in v1.2 we should remove the old implementation, right?

Remove the implementation without removing the API, that will remain and will warn/log it is a noop.

> I don't think it is a good idea to run a VM with passt binding that has nothing behind it and not returning an error? What binding this VM will receive? The user will think all is good and the VM is running but behind the scenes it will receive a binding the user didn't request?

IMO we have no choice, we cannot break existing users, even if they use an alpha or beta feature.
We need to allow their VM/s to operate as close as possible to what they wanted.
I thought about falling back to the "default" or to masquerade binding (which is the closest to passt).

The user should get events about his problematic configuration.
The alternative of suddenly blocking his VM from even running is worse.

But I am open for other opinions in this regard.

Reply:
Declaring a feature as alpha, including alpha fields in graduated APIs, and not being able to revert it, make these deceleration pointless and out life harder, not to mention the API is getting ugly.
We cannot foresee the future and predict what changes will turned out to be irrelevant, we should have a way to do such changes.

Passt its not graduated feature, and as one users who decided using it anyways should know it may break.
As long that we provide a way and enough time for users to prepare for it I thing its fine.

> IMO we have no choice, we cannot break existing users, even if they use an alpha or beta feature.
We need to allow their VM/s to operate as close as possible to what they wanted.
I thought about falling back to the "default" or to masquerade binding (which is the closest to passt).

It can work for passt binding, but its odd, the VM is set with different settings then the desired spec.

Can we call it a temporary state and remove the field from the API entirely at v1.3?

Reply:
> Declaring a feature as alpha, including alpha fields in graduated APIs, and not being able to revert it, make these deceleration pointless and out life harder, not to mention the API is getting ugly.
> We cannot foresee the future and predict what changes will turned out to be irrelevant, we should have a way to do such changes.

This is how k8s have deprecated such fields and it has to do with their persistence in the DB.
In any case, such things are beyond the scope of this specific proposal and can be raised in the sig-api meetings.

> It can work for passt binding, but its odd, the VM is set with different settings then the desired spec.

it is indeed odd, but it does solve the problem of performing an agressive act that breaks existing VMs.
Anyway, as was commented in another thread, we can start with the breaking option and see the feedback.

> Can we call it a temporary state and remove the field from the API entirely at v1.3?

Fields cannot be removed, only the implementation can be removed.